lib/kitchen/element_base.rb

```diff
@@ -858,7 +894,9 @@ def rex_link
       remove_attribute('data-is-for-rex-linking')
 
       unless element_with_ancestors
-        raise "Cannot create rex link to element #{self} - needs ancestors of both types chapter & page/composite_page"
+        raise("Cannot create rex link to element #{self} - needs ancestors of both types chapter & page/composite_page" \
+              "#{data_source ? "\nCNXML SOURCE: " : nil}#{data_source}"
```

Comment:
I'm seeing this "if data source exists, add a newline and sometimes "CNXML SOURCE: ", then append the data source, otherwise add nothing" code repeated in a lot of places. is there a way to memoize? (one idea: an `ElementBase#say_source_or_nil` function)